### PR TITLE
[Model] Implement simple c++ models

### PR DESCRIPTION
**Pull request by hyunsube**
## PR Checklist
Please check if your PR fulfills the following requirements:
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [x] Feature add
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Testing code
- [ ] Other

## What is the new behavior?
- Implement simple c++ model using pytorch c++ and onnx c++.
- Add docs for AI Model

## Other information
- I will compare both ways in terms of inference & training speed.

### CODE REVIEW

File: model/CMakeLists.txt
Comment (sseoreo) [inline model/CMakeLists.txt:1]: I encountered the "fatal error: onnxruntime_cxx_api.h: No such file or directory", so I added some commands to link ONNX libraries like the following: 
```cmake
# Include directories for header files
target_include_directories(net_onnx PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime-linux-x64-1.17.1/include)

# Link the shared library
target_link_libraries(net_onnx PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime-linux-x64-1.17.1/lib/libonnxruntime.so)
```

Also, I met the error "[Undefined reference to symbol 'pthread_create@@GLIBC_2.2.5'](https://stackoverflow.com/questions/34143265/undefined-reference-to-symbol-pthread-createglibc-2-2-5)",
so I set a flag: `SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")`.

In conclusion, my CMakefileLists.txt is modified like following:

```cmake
cmake_minimum_required(VERSION 3.0 FATAL_ERROR)
project(dcgan)


find_package(Torch REQUIRED)

SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")

add_executable(dcgan dcgan.cpp)
add_executable(net_onnx net_onnx.cpp)


target_link_libraries(dcgan "${TORCH_LIBRARIES}")



# Include directories for header files
target_include_directories(net_onnx PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime-linux-x64-1.17.1/include)

# Link the shared library
target_link_libraries(net_onnx PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/onnxruntime-linux-x64-1.17.1/lib/libonnxruntime.so)


set_property(TARGET dcgan PROPERTY CXX_STANDARD 17)
```

File: model/README.md
Comment (sseoreo) [inline model/README.md:1]: When I run the dcgan executable filie, I met the following error:
```
sseo@sseo-PS42-8RB:~/Proj/AI-LINUX-RUST/model/build$ ./dcgan 
terminate called after throwing an instance of 'c10::Error'
  what():  Error opening images file at ./mnist/train-images-idx3-ubyte
Exception raised from read_images at ../torch/csrc/api/src/data/datasets/mnist.cpp:67 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::string) + 0x57 (0x7ffa4bd6bb77 in /home/sseo/Proj/AI-LINUX-RUST/model/build/libtorch/lib/libc10.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::string const&) + 0x64 (0x7ffa4bd1beaf in /home/sseo/Proj/AI-LINUX-RUST/model/build/libtorch/lib/libc10.so)
frame #2: <unknown function> + 0x5b45e75 (0x7ffa3a433e75 in /home/sseo/Proj/AI-LINUX-RUST/model/build/libtorch/lib/libtorch_cpu.so)
frame #3: torch::data::datasets::MNIST::MNIST(std::string const&, torch::data::datasets::MNIST::Mode) + 0x36 (0x7ffa3a434a76 in /home/sseo/Proj/AI-LINUX-RUST/model/build/libtorch/lib/libtorch_cpu.so)
frame #4: main + 0x8c (0x55fe5de1c1b7 in ./dcgan)
frame #5: __libc_start_main + 0xf3 (0x7ffa344ea083 in /lib/x86_64-linux-gnu/libc.so.6)
frame #6: _start + 0x2e (0x55fe5de1bf4e in ./dcgan)
```
I think it is due to MNIST files not found. Did you do something about MNIST?

File: model/CMakeLists.txt
Comment (sseoreo) [reply model/CMakeLists.txt:1]: Could you try this CMakefile in your environment? If it works, then please merge this file.

File: model/train_model_onnx.py
Comment (sseoreo) [inline model/train_model_onnx.py:1]: I can't use cuda device, so I changed `.cuda()` to `.to(device)`.
The corresponding updated code is provided below: 

```python
import torch
import torch.nn as nn
import torch.optim as optim
import torchvision
import torchvision.transforms as transforms
from torch.utils.data import DataLoader

# Define the neural network architecture
class Net(nn.Module):
    def __init__(self):
        super(Net, self).__init__()
        self.fc1 = nn.Linear(784, 64)
        self.fc2 = nn.Linear(64, 32)
        self.fc3 = nn.Linear(32, 10)
        self.dropout = nn.Dropout(0.5)

    def forward(self, x):
        x = torch.relu(self.fc1(x.view(-1, 784)))
        x = self.dropout(x)
        x = torch.relu(self.fc2(x))
        x = torch.log_softmax(self.fc3(x), dim=1)
        return x


if __name__ == '__main__':

    device = "cuda" if torch.cuda.is_available() else "cpu"

    # Load the MNIST dataset
    transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5,), (0.5,))])
    train_dataset = torchvision.datasets.MNIST(root='./build/mnist', train=True, transform=transform, download=True)
    train_loader = DataLoader(train_dataset, batch_size=256, shuffle=True)

    # Initialize the neural network, loss function, and optimizer
    net = Net()
    criterion = nn.NLLLoss()
    optimizer = optim.SGD(net.parameters(), lr=0.01)
    net.train()
    net.to(device) 
    # Train the neural network
    for epoch in range(1, 11):
        running_loss = 0.0
        for i, data in enumerate(train_loader, 0):
            inputs, labels = data
            optimizer.zero_grad()
            outputs = net(inputs.to(device))
            loss = criterion(outputs, labels.to(device))
            loss.backward()
            optimizer.step()
            running_loss += loss.item()
            if i % 100 == 99:    # Print every 100 mini-batches
                print(f'Epoch: {epoch}, Batch: {i+1}, Loss: {running_loss / 100}')
                running_loss = 0.0
    torch.save(net.state_dict(), './net.pt')



```

File: model/convert_to_onnx.py
Comment (sseoreo) [inline model/convert_to_onnx.py:1]: I can't use cuda device, so I changed .cuda() to .to(device).
The corresponding updated code is provided below:

```python

import torch
from train_model_onnx import Net
# Save the trained model

device = 'cuda' if torch.cuda.is_available() else 'cpu'
net = Net().to(device)

net.load_state_dict(torch.load('./net.pt'))
net.eval()

x = torch.randn((1, 1, 28, 28)).to(device)

torch.onnx.export(net,
                  x,
                  "net.onnx",
                  export_params = True,
                  opset_version = 11,
                  do_constant_folding = True,
                  input_names=['input'],
                  output_names=['output'])

```

File: model/dcgan.cpp
Comment (sseoreo) [inline model/dcgan.cpp:1]: Interesting!
It might be useful in future works with DL!

**Review by sseoreo — COMMENTED** [review]
Idk to request changes for my update.
I just leave my comment with my updated codes

File: model/CMakeLists.txt
Comment (hyunsube) [reply model/CMakeLists.txt:1]: Oh, I put those files into `/usr/loca/lib` & `/usr/local/include/` folder and add LD_PATH on ~/,basrc . Let's discuss in next meeting.

File: model/README.md
Comment (hyunsube) [reply model/README.md:1]: Sorry,, if you run the `train_model_onnx.py`, then the MNIST dataset is downloaded automatically. And I found a bug about file path in `dcgan.cpp`. I will change it~

```
  auto data_loader = torch::data::make_data_loader(
      torch::data::datasets::MNIST("./mnist/MNIST/raw").map(
          torch::data::transforms::Stack<>()),
      /*batch_size=*/64);
```

File: model/train_model_onnx.py
Comment (hyunsube) [reply model/train_model_onnx.py:1]: I will apply

File: model/convert_to_onnx.py
Comment (hyunsube) [reply model/convert_to_onnx.py:1]: I will apply!

**Comment by hyunsube** [comment]
@sseoreo Thanks~ I've applied your reviews. If it looks good to you, please merge this PR~
